class_allocation.py

```diff
@@ -299,78 +301,66 @@ def ecpc(self):
-
-                # Summing all future emissions over the future_emissions_timeframe
-                future_emissions_w = self.emis_fut.sel(Time = future_emissions_timeframe).sum(dim='Time')
-
-                total_emissions_w = hist_emissions_w + future_emissions_w
-
-                # Calculating the cumulative population shares for region and world
-                cum_pop = self.xr_total.Population.sel(Time = self.analysis_timeframe).sum(dim='Time')
-                cum_pop_r = cum_pop.sel(Region=self.focus_region)
-                cum_pop_w = cum_pop.sel(Region='EARTH')
-                share_cum_pop = cum_pop_r / cum_pop_w
-                budget_rightful = total_emissions_w * share_cum_pop
-                budget_left = budget_rightful - hist_emissions_r
-
-                # Now temporal allocation
-                #globalbudget = self.xr_total.CO2_globe.sel(Time=self.analysis_timeframe).sum(dim='Time')
-                globalpath = self.emis_fut
-
-                emis_start_r = self.emis_hist.sel(Time=self.start_year_analysis,
-                                                            Region=self.focus_region)
-                emis_start_w = self.emis_hist.sel(Time=self.start_year_analysis,
-                                                            Region='EARTH')
-                emis_ratio = emis_start_r / emis_start_w
-                path_scaled_0 = emis_ratio * globalpath
-                budget_without_assumptions = path_scaled_0.sum(dim='Time')
-                budget_surplus = budget_left - budget_without_assumptions
-
-                def ecpc_factor(initial_path, f):
-                    '''
-                    Calculates a modified emissions path. Takes scaling factor f as input and
-                    returns a new emissions path by scaling the compensation form (xr_comp) with f
-                    and adding it to the initial emissions path (path_scaled_0).
-                    '''
-                    return initial_path + xr_comp * f
-
-                ecpc = ecpc_factor(path_scaled_0, budget_surplus)
-                ecpc_expanded = ecpc.expand_dims(Discount_factor=[discount],
-                                                 Historical_startyear=[startyear]).to_dataset(name='ECPC')
-                xrs.append(ecpc_expanded)
-
-        xr_ecpc = xr.merge(xrs)
-        self.xr_total = self.xr_total.assign(ECPC = xr_ecpc.ECPC)
+                hist_emissions_rt = (hist_emissions * xr_discount)
+                hist_emissions_wt = (hist_emissions * xr_discount).sel(Region='EARTH')
+                Var_H = (share_popt_past*hist_emissions_wt - hist_emissions_rt).sel(Time=np.arange(startyear, 2020+1)).sum(dim='Time').sel(Region=self.focus_region)          # Historical leftover (or debt if negative)
+
+                for conv_year in self.dim_convyears:
+
+                    # Set up starting variables that will change over iteration
+                    Var_E = (global_emissions_future.sel(Time=2021)*GF_frac)
```

Comment:
Better to use `self.start_year_analysis` here and in following lines instead of hardcoding the year 2021.